File: psets/sentimental/readability/readability.py
from cs50 import get_string
import re


def main():
    # 適当な文字列が入力されるまで再帰呼び出し
    string = getstrings()
    
    # !, ? and . is end of a sentense
    sentence = len(re.findall(r'[\!\?\.][\s\0]', string))
    letter = len(re.findall(r'[a-z]', string.lower()))
    words = len(re.findall(r"[a-z][\!\?\.,;:']?[\s\0]", string.lower()))
    # wordsとsentenceは文字列の最後を認識しないため，その分を足した。
    # 全てのテストケースで正常動作して居るが，バグが出る可能性あり。
    cliindex(coleman(letter, words+1, sentence+1))

# coleman(float letter, float s)
# @param letter 100単語あたりの平均文字数
# @param string 100単語あたりの平均文数
# return int    コールマンリアウインデックスを適用した値


def coleman(letters, words, sentence):
    l = letters / words * 100
    s = sentence / words * 100
    return round((0.0588 * l) - (0.296 * s) - 15.8)


# getstring(void)
# return    string
#   適当な値が入力されるまで再帰的に自身を呼び出す。
def getstrings():
    string = get_string("Text: ")
    if string is None:
        return getstrings()
    else:
        return string


# cliindex(int index)
# @param int index  コールマンリアウインデックスが適用された値が入力されると
#                   それに応じたグレードを標準出力する.
def cliindex(index):
    if index >= 16:
        print("Grade 16+")
    elif index <= 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


main()

File: psets/readability/readability.c
// Caleman-Liau-index
#include "cs50.h"
#include "stdio.h"
#include "string.h"
#include "math.h"

int count_letters(char text[]);
int count_words(char text[]);
int count_sentences(char text[]);
int calc_cli(int letters, int words, int senences);
void cli_grade(int index);


int main(void)
{
    int letters, words, sentences, index;
    string text = get_string("Text: ");
    letters = count_letters(text);
    words = count_words(text);
    sentences = count_sentences(text);
    printf("%i letter(s)\n", letters);
    printf("%i word(s)\n", words);
    printf("%i sentence(s)\n", sentences);
    index = calc_cli(letters, words, sentences);
    cli_grade(index);
}


// 文字数の計測
// a-z, A-Zのみを対象として文字数をカウントする。
int count_letters(char text[])
{
    int count = 0;
    for(int i = 0, len = strlen(text); i < len; i++)
    {
        if  ((text[i] >= 'a' && text[i] <= 'z') || (text[i] >= 'A' && text[i] <= 'Z'))
        {
            count++;
        }
    }
    return count;
}

// 単語数の計測
// スペースで区切られた文字を単語"sister-in-law"をみなす。
// ハイフンで繋がれた単語は、まとめて一つの単語として考える。
int count_words(char text[])
{
    int count = 1;
    for(int i = 0, len = strlen(text); i < len; i++)
    {
        if  (text[i] == 32)
        {
            count++;
        }
    }
    return count;
}

// 文章の計測
// '.', '!', '?' を文の終わりとする。
int count_sentences(char text[])
{
    int count = 0;
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if(text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }
    return count;
}

int calc_cli(int letters, int words, int sentences)
{
    float l = (float)letters / (float)words * 100;
    float s = (float)sentences / (float)words * 100;
    float index = 0.0588 * l - 0.296 * s - 15.8;
    return round(index);
}

void cli_grade(int index)
{
    if (index <= 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}